src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/Common/EmbeddedSeparatedSyntaxNodeList.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.Common
+{
+    internal struct EmbeddedSeparatedSyntaxNodeList<TSyntaxKind, TSyntaxNode, TDerivedNode>
+        where TSyntaxKind : struct
+        where TSyntaxNode : EmbeddedSyntaxNode<TSyntaxKind, TSyntaxNode>
+        where TDerivedNode : TSyntaxNode
+    {
+        public ImmutableArray<EmbeddedSyntaxNodeOrToken<TSyntaxKind, TSyntaxNode>> NodesAndTokens { get; }
+        public int Length { get; }
+        public int SeparatorLength { get; }
+
+        public static readonly EmbeddedSeparatedSyntaxNodeList<TSyntaxKind, TSyntaxNode, TDerivedNode> Empty
+            = new(ImmutableArray<EmbeddedSyntaxNodeOrToken<TSyntaxKind, TSyntaxNode>>.Empty);
+
+        public EmbeddedSeparatedSyntaxNodeList(
+            ImmutableArray<EmbeddedSyntaxNodeOrToken<TSyntaxKind, TSyntaxNode>> nodesAndTokens)
+        {
+            Contract.ThrowIfTrue(nodesAndTokens.IsDefault);
+            NodesAndTokens = nodesAndTokens;
+
+            var allLength = NodesAndTokens.Length;
+            Length = (allLength + 1) / 2;
+            SeparatorLength = allLength / 2;
+
+            Verify();
+        }
+
+        [Conditional("DEBUG")]
+        private void Verify()
+        {
+            for (var i = 0; i < NodesAndTokens.Length; i++)
+            {
+                if ((i & 1) == 0)
+                {
+                    // All even values should be TNode
+                    Debug.Assert(NodesAndTokens[i].IsNode);
+                    Debug.Assert(NodesAndTokens[i].Node is EmbeddedSyntaxNode<TSyntaxKind, TSyntaxNode>);
+                }
+                else
+                {
+                    // All odd values should be separator tokens 
+                    Debug.Assert(!NodesAndTokens[i].IsNode);
```

Comment:
can you check that the Kind is not 0 for either the tokens or nodes?

Reply:
How do I compare a `struct` to `int` in this context?

Reply:
ah yeah.  you'd need EmbeddedSyntax to export a raw-kind.  Ignore this for now :)